File: src/platforms/android/app/src/main/java/com/WangWangPhone/core/PersonaCardDbHelper.kt
package com.WangWangPhone.core

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import org.json.JSONArray
import org.json.JSONObject

data class PersonaCard(
    val id: Long = 0,
    val name: String,
    val apiPresetId: Long,
    val createdAt: Long = System.currentTimeMillis(),
    val updatedAt: Long = System.currentTimeMillis()
)

data class PersonaMessage(
    val id: Long = 0,
    val cardId: Long,
    val role: String,
    val content: String,
    val timestamp: Long = System.currentTimeMillis()
)

class PersonaCardDbHelper(context: Context) : SQLiteOpenHelper(context, DATABASE_NAME, null, DATABASE_VERSION) {

    companion object {
        private const val DATABASE_NAME = "persona_cards.db"
        private const val DATABASE_VERSION = 1

        private const val TABLE_CARDS = "persona_cards"
        private const val TABLE_MESSAGES = "persona_messages"

        private const val COLUMN_ID = "id"
        private const val COLUMN_NAME = "name"
        private const val COLUMN_API_PRESET_ID = "api_preset_id"
        private const val COLUMN_CREATED_AT = "created_at"
        private const val COLUMN_UPDATED_AT = "updated_at"

        private const val COLUMN_CARD_ID = "card_id"
        private const val COLUMN_ROLE = "role"
        private const val COLUMN_CONTENT = "content"
        private const val COLUMN_TIMESTAMP = "timestamp"
    }

    override fun onCreate(db: SQLiteDatabase) {
        val createCardsTable = """
            CREATE TABLE $TABLE_CARDS (
                $COLUMN_ID INTEGER PRIMARY KEY AUTOINCREMENT,
                $COLUMN_NAME TEXT NOT NULL,
                $COLUMN_API_PRESET_ID INTEGER NOT NULL,
                $COLUMN_CREATED_AT INTEGER NOT NULL,
                $COLUMN_UPDATED_AT INTEGER NOT NULL
            )
        """.trimIndent()

        val createMessagesTable = """
            CREATE TABLE $TABLE_MESSAGES (
                $COLUMN_ID INTEGER PRIMARY KEY AUTOINCREMENT,
                $COLUMN_CARD_ID INTEGER NOT NULL,
                $COLUMN_ROLE TEXT NOT NULL,
                $COLUMN_CONTENT TEXT NOT NULL,
                $COLUMN_TIMESTAMP INTEGER NOT NULL,
                FOREIGN KEY($COLUMN_CARD_ID) REFERENCES $TABLE_CARDS($COLUMN_ID) ON DELETE CASCADE
            )
        """.trimIndent()

        db.execSQL(createCardsTable)
        db.execSQL(createMessagesTable)
        db.execSQL("CREATE INDEX idx_messages_card_id ON $TABLE_MESSAGES($COLUMN_CARD_ID)")
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        db.execSQL("DROP TABLE IF EXISTS $TABLE_MESSAGES")
        db.execSQL("DROP TABLE IF EXISTS $TABLE_CARDS")
        onCreate(db)
    }

    fun createCard(name: String, apiPresetId: Long): Long {
        val db = writableDatabase
        val values = ContentValues().apply {
            put(COLUMN_NAME, name)
            put(COLUMN_API_PRESET_ID, apiPresetId)
            put(COLUMN_CREATED_AT, System.currentTimeMillis())
            put(COLUMN_UPDATED_AT, System.currentTimeMillis())
        }
        return db.insert(TABLE_CARDS, null, values)
    }

    fun getAllCards(): List<PersonaCard> {
        val cards = mutableListOf<PersonaCard>()
        val db = readableDatabase
        val cursor = db.query(
            TABLE_CARDS,
            null,
            null,
            null,
            null,
            null,
            "$COLUMN_UPDATED_AT DESC"
        )

        cursor.use {
            while (it.moveToNext()) {
                cards.add(
                    PersonaCard(
                        id = it.getLong(it.getColumnIndexOrThrow(COLUMN_ID)),
                        name = it.getString(it.getColumnIndexOrThrow(COLUMN_NAME)),
                        apiPresetId = it.getLong(it.getColumnIndexOrThrow(COLUMN_API_PRESET_ID)),
                        createdAt = it.getLong(it.getColumnIndexOrThrow(COLUMN_CREATED_AT)),
                        updatedAt = it.getLong(it.getColumnIndexOrThrow(COLUMN_UPDATED_AT))
                    )
                )
            }
        }
        return cards
    }

    fun getCard(cardId: Long): PersonaCard? {
        val db = readableDatabase
        val cursor = db.query(
            TABLE_CARDS,
            null,
            "$COLUMN_ID = ?",
            arrayOf(cardId.toString()),
            null,
            null,
            null
        )

        cursor.use {
            if (it.moveToFirst()) {
                return PersonaCard(
                    id = it.getLong(it.getColumnIndexOrThrow(COLUMN_ID)),
                    name = it.getString(it.getColumnIndexOrThrow(COLUMN_NAME)),
                    apiPresetId = it.getLong(it.getColumnIndexOrThrow(COLUMN_API_PRESET_ID)),
                    createdAt = it.getLong(it.getColumnIndexOrThrow(COLUMN_CREATED_AT)),
                    updatedAt = it.getLong(it.getColumnIndexOrThrow(COLUMN_UPDATED_AT))
                )
            }
        }
        return null
    }

    fun updateCard(cardId: Long, name: String) {
        val db = writableDatabase
        val values = ContentValues().apply {
            put(COLUMN_NAME, name)
            put(COLUMN_UPDATED_AT, System.currentTimeMillis())
        }
        db.update(TABLE_CARDS, values, "$COLUMN_ID = ?", arrayOf(cardId.toString()))
    }

    fun deleteCard(cardId: Long) {
        val db = writableDatabase
        db.delete(TABLE_CARDS, "$COLUMN_ID = ?", arrayOf(cardId.toString()))
    }

    fun saveMessage(cardId: Long, role: String, content: String): Long {
        val db = writableDatabase
        val values = ContentValues().apply {
            put(COLUMN_CARD_ID, cardId)
            put(COLUMN_ROLE, role)
            put(COLUMN_CONTENT, content)
            put(COLUMN_TIMESTAMP, System.currentTimeMillis())
        }
        
        // 更新卡片的更新时间
        val cardValues = ContentValues().apply {
            put(COLUMN_UPDATED_AT, System.currentTimeMillis())
        }
        db.update(TABLE_CARDS, cardValues, "$COLUMN_ID = ?", arrayOf(cardId.toString()))
        
        return db.insert(TABLE_MESSAGES, null, values)
    }

    fun getMessages(cardId: Long): List<PersonaMessage> {
        val messages = mutableListOf<PersonaMessage>()
        val db = readableDatabase
        val cursor = db.query(
            TABLE_MESSAGES,
            null,
            "$COLUMN_CARD_ID = ?",
            arrayOf(cardId.toString()),
            null,
            null,
            "$COLUMN_TIMESTAMP ASC"
        )

        cursor.use {
            while (it.moveToNext()) {
                messages.add(
                    PersonaMessage(
                        id = it.getLong(it.getColumnIndexOrThrow(COLUMN_ID)),
                        cardId = it.getLong(it.getColumnIndexOrThrow(COLUMN_CARD_ID)),
                        role = it.getString(it.getColumnIndexOrThrow(COLUMN_ROLE)),
                        content = it.getString(it.getColumnIndexOrThrow(COLUMN_CONTENT)),
                        timestamp = it.getLong(it.getColumnIndexOrThrow(COLUMN_TIMESTAMP))
                    )
                )
            }
        }
        return messages
    }

    fun clearMessages(cardId: Long) {
        val db = writableDatabase
        db.delete(TABLE_MESSAGES, "$COLUMN_CARD_ID = ?", arrayOf(cardId.toString()))
    }
}

File: src/platforms/ios/WangWangPhone/Core/PersonaCardDbHelper.swift
import Foundation
import SQLite3

struct PersonaCard: Identifiable {
    let id: Int64
    let name: String
    let apiPresetId: Int64
    let createdAt: Int64
    let updatedAt: Int64
}

struct PersonaMessage {
    let id: Int64
    let cardId: Int64
    let role: String
    let content: String
    let timestamp: Int64
}

class PersonaCardDbHelper {
    private var db: OpaquePointer?
    
    init() {
        let fileURL = try! FileManager.default
            .url(for: .documentDirectory, in: .userDomainMask, appropriateFor: nil, create: false)
            .appendingPathComponent("WangWangPhone.db")
        
        if sqlite3_open(fileURL.path, &db) != SQLITE_OK {
            print("Error opening database")
        }
        
        createTables()
    }
    
    deinit {
        sqlite3_close(db)
    }
    
    private func createTables() {
        let createCardTable = """
        CREATE TABLE IF NOT EXISTS persona_cards (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            name TEXT NOT NULL,
            api_preset_id INTEGER NOT NULL,
            created_at INTEGER NOT NULL,
            updated_at INTEGER NOT NULL
        )
        """
        
        let createMessageTable = """
        CREATE TABLE IF NOT EXISTS persona_messages (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            card_id INTEGER NOT NULL,
            role TEXT NOT NULL,
            content TEXT NOT NULL,
            timestamp INTEGER NOT NULL,
            FOREIGN KEY (card_id) REFERENCES persona_cards(id) ON DELETE CASCADE
        )
        """
        
        if sqlite3_exec(db, createCardTable, nil, nil, nil) != SQLITE_OK {
            print("Error creating persona_cards table")
        }
        
        if sqlite3_exec(db, createMessageTable, nil, nil, nil) != SQLITE_OK {
            print("Error creating persona_messages table")
        }
    }
    
    func createCard(name: String, apiPresetId: Int64) -> Int64 {
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        let query = "INSERT INTO persona_cards (name, api_preset_id, created_at, updated_at) VALUES (?, ?, ?, ?)"
        var statement: OpaquePointer?
        
        if sqlite3_prepare_v2(db, query, -1, &statement, nil) == SQLITE_OK {
            sqlite3_bind_text(statement, 1, (name as NSString).utf8String, -1, nil)
            sqlite3_bind_int64(statement, 2, apiPresetId)
            sqlite3_bind_int64(statement, 3, now)
            sqlite3_bind_int64(statement, 4, now)
            
            if sqlite3_step(statement) == SQLITE_DONE {
                let id = sqlite3_last_insert_rowid(db)
                sqlite3_finalize(statement)
                return id
            }
        }
        
        sqlite3_finalize(statement)
        return -1
    }
    
    func getAllCards() -> [PersonaCard] {
        let query = "SELECT id, name, api_preset_id, created_at, updated_at FROM persona_cards ORDER BY updated_at DESC"
        var statement: OpaquePointer?
        var cards: [PersonaCard] = []
        
        if sqlite3_prepare_v2(db, query, -1, &statement, nil) == SQLITE_OK {
            while sqlite3_step(statement) == SQLITE_ROW {
                let id = sqlite3_column_int64(statement, 0)
                let name = String(cString: sqlite3_column_text(statement, 1))
                let apiPresetId = sqlite3_column_int64(statement, 2)
                let createdAt = sqlite3_column_int64(statement, 3)
                let updatedAt = sqlite3_column_int64(statement, 4)
                
                cards.append(PersonaCard(
                    id: id,
                    name: name,
                    apiPresetId: apiPresetId,
                    createdAt: createdAt,
                    updatedAt: updatedAt
                ))
            }
        }
        
        sqlite3_finalize(statement)
        return cards
    }
    
    func getCardById(_ id: Int64) -> PersonaCard? {
        let query = "SELECT id, name, api_preset_id, created_at, updated_at FROM persona_cards WHERE id = ?"
        var statement: OpaquePointer?
        
        if sqlite3_prepare_v2(db, query, -1, &statement, nil) == SQLITE_OK {
            sqlite3_bind_int64(statement, 1, id)
            
            if sqlite3_step(statement) == SQLITE_ROW {
                let id = sqlite3_column_int64(statement, 0)
                let name = String(cString: sqlite3_column_text(statement, 1))
                let apiPresetId = sqlite3_column_int64(statement, 2)
                let createdAt = sqlite3_column_int64(statement, 3)
                let updatedAt = sqlite3_column_int64(statement, 4)
                
                sqlite3_finalize(statement)
                return PersonaCard(
                    id: id,
                    name: name,
                    apiPresetId: apiPresetId,
                    createdAt: createdAt,
                    updatedAt: updatedAt
                )
            }
        }
        
        sqlite3_finalize(statement)
        return nil
    }
    
    func deleteCard(_ id: Int64) {
        let query = "DELETE FROM persona_cards WHERE id = ?"
        var statement: OpaquePointer?
        
        if sqlite3_prepare_v2(db, query, -1, &statement, nil) == SQLITE_OK {
            sqlite3_bind_int64(statement, 1, id)
            sqlite3_step(statement)
        }
        
        sqlite3_finalize(statement)
    }
    
    func addMessage(_ message: PersonaMessage) -> Int64 {
        let query = "INSERT INTO persona_messages (card_id, role, content, timestamp) VALUES (?, ?, ?, ?)"
        var statement: OpaquePointer?
        
        if sqlite3_prepare_v2(db, query, -1, &statement, nil) == SQLITE_OK {
            sqlite3_bind_int64(statement, 1, message.cardId)
            sqlite3_bind_text(statement, 2, (message.role as NSString).utf8String, -1, nil)
            sqlite3_bind_text(statement, 3, (message.content as NSString).utf8String, -1, nil)
            sqlite3_bind_int64(statement, 4, message.timestamp)
            
            if sqlite3_step(statement) == SQLITE_DONE {
                let id = sqlite3_last_insert_rowid(db)
                sqlite3_finalize(statement)
                return id
            }
        }
        
        sqlite3_finalize(statement)
        return -1
    }
    
    func getMessages(cardId: Int64) -> [PersonaMessage] {
        let query = "SELECT id, card_id, role, content, timestamp FROM persona_messages WHERE card_id = ? ORDER BY timestamp ASC"
        var statement: OpaquePointer?
        var messages: [PersonaMessage] = []
        
        if sqlite3_prepare_v2(db, query, -1, &statement, nil) == SQLITE_OK {
            sqlite3_bind_int64(statement, 1, cardId)
            
            while sqlite3_step(statement) == SQLITE_ROW {
                let id = sqlite3_column_int64(statement, 0)
                let cardId = sqlite3_column_int64(statement, 1)
                let role = String(cString: sqlite3_column_text(statement, 2))
                let content = String(cString: sqlite3_column_text(statement, 3))
                let timestamp = sqlite3_column_int64(statement, 4)
                
                messages.append(PersonaMessage(
                    id: id,
                    cardId: cardId,
                    role: role,
                    content: content,
                    timestamp: timestamp
                ))
            }
        }
        
        sqlite3_finalize(statement)
        return messages
    }
    
    func updateCardTimestamp(_ id: Int64) {
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        let query = "UPDATE persona_cards SET updated_at = ? WHERE id = ?"
        var statement: OpaquePointer?
        
        if sqlite3_prepare_v2(db, query, -1, &statement, nil) == SQLITE_OK {
            sqlite3_bind_int64(statement, 1, now)
            sqlite3_bind_int64(statement, 2, id)
            sqlite3_step(statement)
        }
        
        sqlite3_finalize(statement)
    }
    
    func updateMessageContent(cardId: Int64, messageId: Int64, newContent: String) -> Bool {
        let query = "UPDATE persona_messages SET content = ? WHERE card_id = ? AND id = ?"
        var statement: OpaquePointer?
        
        if sqlite3_prepare_v2(db, query, -1, &statement, nil) == SQLITE_OK {
            sqlite3_bind_text(statement, 1, (newContent as NSString).utf8String, -1, nil)
            sqlite3_bind_int64(statement, 2, cardId)
            sqlite3_bind_int64(statement, 3, messageId)
            
            let result = sqlite3_step(statement) == SQLITE_DONE
            sqlite3_finalize(statement)
            return result
        }
        
        sqlite3_finalize(statement)
        return false
    }
    
    func deleteMessagesAfter(cardId: Int64, messageId: Int64) {
        let query = "DELETE FROM persona_messages WHERE card_id = ? AND id > ?"
        var statement: OpaquePointer?
        
        if sqlite3_prepare_v2(db, query, -1, &statement, nil) == SQLITE_OK {
            sqlite3_bind_int64(statement, 1, cardId)
            sqlite3_bind_int64(statement, 2, messageId)
            sqlite3_step(statement)
        }
        
        sqlite3_finalize(statement)
    }
    
    /// 更新人设卡（对齐 Android updateCard）
    func updateCard(id: Int64, name: String, apiPresetId: Int64) -> Bool {
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        let query = "UPDATE persona_cards SET name = ?, api_preset_id = ?, updated_at = ? WHERE id = ?"
        var statement: OpaquePointer?
        
        if sqlite3_prepare_v2(db, query, -1, &statement, nil) == SQLITE_OK {
            sqlite3_bind_text(statement, 1, (name as NSString).utf8String, -1, nil)
            sqlite3_bind_int64(statement, 2, apiPresetId)
            sqlite3_bind_int64(statement, 3, now)
            sqlite3_bind_int64(statement, 4, id)
            
            let result = sqlite3_step(statement) == SQLITE_DONE
            sqlite3_finalize(statement)
            return result
        }
        
        sqlite3_finalize(statement)
        return false
    }
    
    /// 清空某张卡的所有消息（对齐 Android clearMessages）
    func clearMessages(cardId: Int64) -> Bool {
        let query = "DELETE FROM persona_messages WHERE card_id = ?"
        var statement: OpaquePointer?
        
        if sqlite3_prepare_v2(db, query, -1, &statement, nil) == SQLITE_OK {
            sqlite3_bind_int64(statement, 1, cardId)
            let result = sqlite3_step(statement) == SQLITE_DONE
            sqlite3_finalize(statement)
            return result
        }
        
        sqlite3_finalize(statement)
        return false
    }
}
